### PR TITLE
terraform/hydra-nixpkgs: add unfree-redist

### PR DESCRIPTION
**Pull request by zowoq**

### CODE REVIEW

File: terraform/hydra-projects.tf [hunk 1]
```diff
@@ -170,3 +170,20 @@ resource "hydra_jobset" "nixpkgs_rocm" {
...
+  state       = "enabled"
+  visible     = true
+  name        = "main"
+  type        = "flake"
+  description = "unstable"
+
+  flake_uri = "https://github.com/numtide/nixpkgs-unfree/nixpkgs-unstable"
+
+  check_interval    = 1800
+  scheduling_shares = 5000
```
Comment (zowoq) [inline terraform/hydra-projects.tf:185]: This is 1000 less than cuda/rocm have.

File: terraform/hydra-projects.tf [hunk 2]
```diff
@@ -170,3 +170,20 @@ resource "hydra_jobset" "nixpkgs_rocm" {
   email_notifications = false
   email_override      = ""
 }
+
+resource "hydra_jobset" "nixpkgs_unfree" {
+  project     = hydra_project.nixpkgs.name
+  state       = "enabled"
+  visible     = true
+  name        = "main"
```
Comment (zimbatm) [inline terraform/hydra-projects.tf:178]:
```suggestion
  name        = "unfree"
```
?

**Review by zimbatm — APPROVED** [review]
LGTM

**Comment by zowoq** [comment]
https://hydra.nix-community.org/jobset/nixpkgs/unfree

**Comment by zowoq** [comment]
https://hydra.nix-community.org/eval/28213#tabs-removed

`requireFile` is probably about as far as filtering can go, anything further needs to be done by updating the packages in nixpkgs:

https://github.com/NixOS/nixpkgs/pull/335271
https://hydra.nix-community.org/eval/33493#tabs-removed

___

I think the relevant parts of numtide/nixpkgs-unfree can upstreamed to nixpkgs but what I want to do is keep the current filters as the default version that we can keep building frequently and a "full" version behind an option that can be used for the code coverage job but on a slower channel, built less frequently and at a lower priority than the other hydra jobsets.

e.g.
```nix
  default = isUnfree && isSource && isNotCudaPackage && isNotLinuxKernel && canSubstituteSrc;

  full = isUnfree && isSource && canSubstituteSrc;
```

```hcl
resource "hydra_jobset" "nixpkgs_unfree_full" {
  nix_expression {
    file  = "pkgs/top-level/release-unfree.nix"
    input = "nixpkgs"
  }

  input {
    name              = "full"
    type              = "boolean"
    value             = "true"
  }

  input {
    name              = "nixpkgs"
    type              = "git"
    value             = "https://github.com/NixOS/nixpkgs.git nixos-unstable"
  }

  check_interval    = ? # once a week
  scheduling_shares = 1000
}
```

**Comment by zowoq** [comment]
@zimbatm have you seen my previous comment here?

**Comment by zowoq** [comment]
@zimbatm are you going to upstream this?

**Comment by zowoq** [comment]
Cloudflare has made some changes to its provider so I've had to manually downgrade to an older version for now.

unfree: https://hydra.nix-community.org/eval/70021

unfree full: https://hydra.nix-community.org/eval/70023

There are cuda and kernel packages in the `unfree` jobset?

**Comment by zimbatm** [comment]
yes, wasn't that the logic you wanted? https://github.com/NixOS/nixpkgs/pull/342529/files#diff-364739045ddca41a05dac92fc0ff40be35d86f8001d3a66076b3f93cab9dbdf1R89-R100

**Comment by zowoq** [comment]
> yes, wasn't that the logic you wanted?

Currently they are in both jobsets, they shouldn't be in `unfree`?

**Comment by zimbatm** [comment]
Ok, I can filter for isNotCudaPackage on the full and non-full jobsets.

**Comment by zimbatm** [comment]
see b59601bc5831823afafb75bff1d39146a84b8695

**Comment by zowoq** [comment]
Sorry, no, that wasn't what I meant, I'll try explaining it again:

There are cuda and linux packages in both jobsets, `unfree` and `unfree-full` but they should only be in `unfree-full`.

`unfree`: https://hydra.nix-community.org/eval/70021

`unfree-full`: https://hydra.nix-community.org/eval/70023

**Comment by zimbatm** [comment]
Ok, I got it now. Fixed the logic in 0632ba36d5fceeef31b1600a9ef8ed7eb6379f3d

**Comment by zowoq** [comment]
Looks good now, thanks.

`unfree`: new eval, cuda and linux removed: https://hydra.nix-community.org/eval/70585 

`unfree-full`: same as the previous eval: https://hydra.nix-community.org/jobset/nixpkgs/unfree-full

**Comment by emilazy** [comment]
This is uploading non‐redistributable binaries to Cachix and illegally redistributing them to users, cf. https://github.com/NixOS/nixpkgs/pull/342529#issuecomment-2358438965. (Apparently Hydra may also offer a direct download that would also be problematic but it’s currently conveniently broken by a Perl exception.)

**Comment by zimbatm** [comment]
@zowoq I'm trying this again with unfree+redistributable. This is more in line with what I believe upstream should be building.

**Comment by zowoq** [comment]
https://hydra.nix-community.org/jobset/nixpkgs/unfree-redist

https://hydra.nix-community.org/jobset/nixpkgs/unfree-redist-full

**Comment by zimbatm** [comment]
This is ready to go

**Comment by zowoq** [comment]
Hydra seems slower than usual, only just landed in nixos-unstable. https://hydra.nix-community.org/eval/120998

Do you want to add an unfree-redist-stable jobset starting with 24.11? There doesn't seem to be much churn on stable (apart from the occasional staging merge) so it should be cheap for us to support.

**Comment by zimbatm** [comment]
> Do you want to add an unfree-redist-stable jobset starting with 24.11? There doesn't seem to be much churn on stable (apart from the occasional staging merge) so it should be cheap for us to support.

yes I think it would be useful

**Comment by sedlund** [comment]
can [this ](https://github.com/nix-community/infra/blob/85e5633d401fd9bcfff05cfc00e8978c19e44090/terraform/hydra-nixpkgs.tf#L63)happen more than once a week?  maybe per day?

(interested in NVidia kernel drivers)

**Comment by zowoq** [comment]
@sedlund https://github.com/nix-community/infra/pull/1863.

**Pull request merged**